Documentation/ceph-cluster-crd.md

```diff
@@ -616,6 +616,73 @@ healthCheck:
 
 Changing the liveness probe is an advanced operation and should rarely be necessary. If you want to change these settings then modify the desired settings.
 
+## Status
+
+The operator is regularly configuring and checking the health of the cluster. The results of the configuration
+and health checks can be seen in the `status` section of the CephCluster CR.
+
+```yaml
+  status:
+    ceph:
+      health: HEALTH_OK
+      lastChecked: "2021-03-02T21:22:11Z"
+      capacity:
+        bytesAvailable: 22530293760
+        bytesTotal: 25757220864
+        bytesUsed: 3226927104
+        lastUpdated: "2021-03-02T21:22:11Z"
+    message: Cluster created successfully
+    phase: Ready
+    state: Created
+    storage:
+      deviceClasses:
+      - name: hdd
+    version:
+      image: ceph/ceph:v15
+      version: 15.2.9-0
+    conditions:
+    - lastHeartbeatTime: "2021-03-02T21:22:11Z"
+      lastTransitionTime: "2021-03-02T21:21:09Z"
+      message: Cluster created successfully
+      reason: ClusterCreated
+      status: "True"
+      type: Ready
+```
+
+### Ceph Status
+
+Ceph is constantly monitoring the health of the data plane and reporting back if there are
+any warnings or errors. If everything is healthy from Ceph's perspective, you will see
+`HEALTH_OK`.
+
+If Ceph reports any warnings or errors, the details will be printed to the status.
+If further troubleshooting is needed to resolve these issues, the toolbox will likely
+be needed where you can run `ceph` commands to find more details.
+
+The `capacity` of the cluster is reported, including bytes available, total, and used.
+The available space will be less that you may expect due to overhead in the OSDs.
+
+### Conditions
+
+The `conditions` represent the status of the Rook operator.
+- If the cluster is fully configured and the operator is stable, the
+  `ClusterCreated` condition is raised with `type: Ready` and no other conditions.
+- If the cluster is externally connected successfully, the only condition will be `ClusterConnected`.
+- If the operator is currently being configured or the operator is checking for update,
+  there will be a `ClusterProgressing` condition.
+- If there was a failure, the condition(s) status will be `false` and the `message` will
+  give a summary of the error. See the operator log for more details.
```

Comment:
I think we should refer to how k8s documents its conditions to be closer to what users will expect. I don't really interpret the `reason` as being the actual condition. The condition is actually the `type` as implied by the k8s nodes condition docs, not the `reason` as our code and docs seem to treat it.

https://kubernetes.io/docs/concepts/architecture/nodes/#condition

IMO, we always need to have the `type: Ready` condition, but the status will be `"False"` until the cluster is created, and we must set it back to `"False"` if the cluster falls into a non-ready state. The `reason` for the cluster condition should be `ClusterCreated` or `ClusterConnected` depending on whether it is an external cluster or not.

Having said this, I wonder if we should have a persistent `type: ClusterProgressing` condition that can be true/false depending on if the  operator is running a reconcile or not.

Reply:
> I think we should refer to how k8s documents its conditions to be closer to what users will expect. I don't really interpret the `reason` as being the actual condition. The condition is actually the `type` as implied by the k8s nodes condition docs, not the `reason` as our code and docs seem to treat it.
> 
> https://kubernetes.io/docs/concepts/architecture/nodes/#condition

There is a predictable mapping between the condition and the reason so they are in reality somewhat interchangeable, but agreed that the focus should be on the type. I'll update the doc. 

> IMO, we always need to have the `type: Ready` condition, but the status will be `"False"` until the cluster is created, and we must set it back to `"False"` if the cluster falls into a non-ready state. The `reason` for the cluster condition should be `ClusterCreated` or `ClusterConnected` depending on whether it is an external cluster or not.

Whether the cluster is actually ready is a Ceph question. Rather than attempt to reflect the Ceph readiness in the conditions, the approach is to defer to the Ceph health status if that question needs to be answered. So the `Ready` condition is just to indicate if the cluster has completed initial reconcile to indicate if it's expected the storage can now be consumed. We expect the cluster to always be in a consumable state after that point unless the Ceph health status is not ok, in which case we refer back to the Ceph status instead of the conditions. Rolling the Ceph status into the conditions is just too complex IMO. I believe consumers such as OCS are also relying on the `Ready` status to indicate that operations dependent on the basic reconcile can be continued.

> Having said this, I wonder if we should have a persistent `type: ClusterProgressing` condition that can be true/false depending on if the operator is running a reconcile or not. (Aside: `ClusterProgressing` might not be a great name either... `ClusterReconciling` could be a little clearer, though the message we put on the condition can help clear up any misconceptions if we want to keep that for backwards compatibility)

Having a `ClusterProgressing: false` was actually the issue that spurred this whole change. Any `false` condition was causing Rancher to interpret it as an error condition. Therefore, if the cluster is not progressing, we now remove that condition and `false` conditions are reserved for error conditions.

Reply:
> Having a `ClusterProgressing: false` was actually the issue that spurred this whole change. Any `false` condition was causing Rancher to interpret it as an error condition. Therefore, if the cluster is not progressing, we now remove that condition and `false` conditions are reserved for error conditions.

Really!? But how does Rancher handle the Kubernetes node conditions like "DiskPressure" "False"?!?!

Reply:
To further prove my point, the k8s sig-architecture api conventions do not suggest that non-True conditions should be treated as errors. Instead, they should have "abnormal-true" polarity, which suggests the opposite of what Rancher is doing.

https://github.com/kubernetes/community/blob/a2cdce5/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties